### PR TITLE
Introducing the new API wrapper

### LINKED ISSUE #546: A common JavaScript toolkit for the front end
It isn't efficient to have each theme define their own way to interact with the back end API.

We should introduce a toolkit with the barebones to interact with the API. It's hard to update every theme's JavaScript code when we introduce something new or breaking. A common toolkit would be easier to maintenance in the long run, and would make it easier for custom themes to appear.

### PR DESCRIPTION
**Pull request by evrifaessa**
This pull request implements the JavaScript wrapper I've created to make it easier to call the FOSSBilling API. The wrapper includes methods to call the admin, client and the guest API.

The wrapper itself is very easy to use, and after we update the documentation, should be very helpful for potential contributors to use.

Here is an example:

```js
API.guest.get("system/version", {}, function(response) {
	console.log(response);
});
```
The method accepts a URL (or a relative path), parameters, a success handler and an error handler.

-----

In addition to making it easier for outside collaborators to create new themes and modules that interact with the backend API, this wrapper also eliminates the need to manually assign the CSRF token with each request.

To achieve this, I've attached the CSRF token into a meta tag in each theme.

```html
 <meta name="csrf-token" content="{{ CSRFToken }}">
```

We can then access the token using the new method. The wrapper automatically looks for that meta tag using this method and attaches it to every request.
image

Once we make sure it works as intended, we can go ahead and remove all the hidden input tags from our forms.

I have removed a fair chunk of jQuery and AJAX code and replaced them with vanilla JS alternatives like the native [fetch API](https://developer.mozilla.org/en-US/docs/Web/API/fetch). This pull request brings us closer to our goal of eventually dropping jQuery entirely.

-----

I've kept things backwards-compatible. This PR implements the new wrapper to the admin theme. It also keeps the legacy `bb.post()` and `bb.get()` functions working with a deprecation error, we can delete them in the far future but for now they should work as usual. They now use the new wrapper, but keep the way you call the function the same.

I haven't even touched Huraga other than just importing the wrapper. I don't want to get the PR bigger.

Please review carefully.

Fixes #546.

### CODE REVIEW

**Comment by BelleNottelling** [comment]
Can we get the updated to remove the workaround for https://github.com/FOSSBilling/FOSSBilling/issues/597?

File: src/themes/admin_default/assets/js/fossbilling.js [hunk 1]
```diff
@@ -85,85 +85,114 @@
...
+    if (obj.classList.contains('bb-rm-tr')) {
+      obj.closest('tr').classList.add('highlight');
       return;
     }
 
-    if (obj.attr('data-api-reload') !== undefined) {
-      this.reload();
-
+    if (obj.hasAttribute('data-api-redirect')) {
+      window.location = obj.getAttribute('data-api-redirect');
```
[the remark this answers is not in the document]
Comment (evrifaessa) [reply src/themes/admin_default/assets/js/fossbilling.js:102]: I'm quite unsure about this one. The `data-api-redirect` attribute comes directly from the form. I can't see how an attacker can exploit that as the data isn't read from the query string or anything. It comes directly from the forms we create.

[the remark this answers is not in the document]
Comment (BelleNottelling) [reply src/themes/admin_default/assets/js/fossbilling.js:102]: I think we can dismiss this one

**Comment by evrifaessa** [comment]
> Can we get the updated to remove the workaround for #597?

@BelleNottelling - Done.

**Review by jaapmarcus — APPROVED** [review]

**Review by BelleNottelling — APPROVED** [review]

**Pull request merged**